File: BOJ/boj_2573_iceberg.java
import java.util.*;
import java.io.*;

class Area {
    int i;
    int j;
    int cntZero;

    Area(int i, int j) {
        this.i = i;
        this.j = j;
    }

    Area(int i, int j, int cntZero) {
        this.i = i;
        this.j = j;
        this.cntZero = cntZero;
    }
}

public class boj_2573_iceberg {
    static int row;
    static int column;
    static int year;
    static int[][] ground;
    static boolean[][] visited;
    static int[] di = {-1, 1, 0, 0};
    static int[] dj = {0, 0, -1, 1};
    public static void main(String[] args) throws IOException {
        // String txtPath = "../input.txt";
        // BufferedReader br = new BufferedReader(new FileReader(txtPath));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] index = br.readLine().split(" ");
        row = Integer.parseInt(index[0]);
        column = Integer.parseInt(index[1]);
        ground = new int[row][column];
        visited = new boolean[row][column];
        year = 0;

        for (int i = 0; i < row; i++) {
            String[] line = br.readLine().split(" ");
            for (int j = 0 ; j < column; j++) {
                ground[i][j] = Integer.parseInt(line[j]);
            }
        }

        while (true) {
            int islandNum = 0;
            resetVistied();
            melting();
            year += 1;            
            for (int i = 0 ; i < row ; i++) {
                for (int j = 0 ; j < column; j++) {
                    if (ground[i][j] != 0 && visited[i][j] == false) {
                        dfs(i, j);
                        islandNum += 1;
                    }
                }
            }
            if (islandNum > 1) {
                break;
            } else if (islandNum == 0) {
                year = 0;
                break;
            }
        }
        System.out.println(year);
        br.close();
    }

    public static boolean isField(int i, int j) {
        boolean res = false;
        if (i >= 0 && i < row && j >= 0 && j < column) {
            res = true;
        }
        return res;
    }

    public static void dfs(int i, int j) {
        Stack<Area> stack = new Stack<>();
        stack.add(new Area(i, j));

        while (!stack.isEmpty()) {
            Area area = stack.pop();
            int si = area.i;
            int sj = area.j;
            for (int k = 0; k < 4; k++) {
                int ni = si + di[k];
                int nj = sj + dj[k];
                if (!isField(ni, nj)) {
                    continue;
                }
                if (ground[ni][nj] != 0 && visited[ni][nj] == false) {
                    stack.add(new Area(ni, nj));
                    visited[ni][nj] = true;
                }
            }
        }
    }

    public static void resetVistied() {
        for (int i = 0 ; i < row ; i++) {
            for (int j = 0 ; j < column; j++) {
                visited[i][j] = false;
            }
        }
    }

    public static void melting() {
        ArrayList<Area> info = new ArrayList<>();
        for (int i = 0 ; i < row ; i++) {
            for (int j = 0 ; j < column; j++) {
                int cntZero = 0;
                if (ground[i][j] == 0) {
                    continue;
                }
                for (int k = 0; k < 4; k++) {
                    int ni = i + di[k];
                    int nj = j + dj[k];
                    if (!isField(ni, nj)) {
                        continue;
                    }
                    if (ground[ni][nj] == 0) {
                        cntZero += 1;
                    }
                }
                info.add(new Area(i, j, cntZero));
            }
        }

        for (int i=0; i < info.size(); i++) {
            Area area = info.get(i);
            if (area.cntZero >= ground[area.i][area.j]) {
                ground[area.i][area.j] = 0;
            } else {
                ground[area.i][area.j] -= area.cntZero;
            }
        }
    }
}

File: BOJ/boj_2573_iceberg.py
import sys
sys.stdin = open('../input.txt', 'r')

ds = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def dfs(i, j):
    stack = [(i, j)]
    visited[i][j] = 1

    while stack:
        si, sj = stack.pop()
        for di, dj in ds:
            ni, nj = si+di, sj+dj
            if not (0 <= ni < N and 0 <= nj < M): continue
            if g[ni][nj] and not visited[ni][nj]:
                stack.append((ni, nj))
                visited[ni][nj] = 1
    return 1


def melt_iceberg():
    infos = []
    for i in range(N):
        for j in range(M):
            sea_count = 0
            if g[i][j] == 0:
                continue
            for di, dj in ds:
                ni, nj = i+di, j+dj
                if g[ni][nj] == 0:
                    sea_count += 1
            infos.append((i, j, sea_count))

    for info in infos:
        i, j, sea_count = info
        g[i][j] = 0 if g[i][j] <= sea_count else g[i][j] - sea_count


N, M = map(int, input().split())
g = [list(map(int, input().split())) for _ in range(N)]
year = 0

while True:
    visited = [[0 for _ in range(M)] for _ in range(N)]
    melt_iceberg()
    iceberg = 0
    for i in range(N):
        for j in range(M):
            if g[i][j] != 0 and not visited[i][j]:
                iceberg += dfs(i, j)

    year += 1
    if iceberg > 1:
        break
    elif iceberg == 0:
        year = 0
        break

print(year)